File: examples/java-plain/src/main/java/example/DeployNotification.java
package example;

import java.time.Instant;

import io.github.teams4j.cards.Colors;
import io.github.teams4j.cards.FontSize;
import io.github.teams4j.cards.FontWeight;
import io.github.teams4j.cards.dsl.Cards;
import io.github.teams4j.webhook.WebhookException;
import io.github.teams4j.webhook.WebhookResponse;
import io.github.teams4j.webhook.WorkflowsWebhookClient;

/**
 * A deploy-failure notification with no framework at all.
 *
 * <pre>
 *   export TEAMS_WEBHOOK_URL='https://...'
 *   ./gradlew :java-plain:run
 * </pre>
 *
 * <p>Two dependencies, and one of them is only there to choose a JSON binding. HTTP is the JDK's
 * {@code HttpClient} and logging is {@code System.Logger}, so nothing else comes along.
 */
public final class DeployNotification {

    public static void main(String[] args) {
        String url = System.getenv("TEAMS_WEBHOOK_URL");
        if (url == null || url.isBlank()) {
            System.err.println("set TEAMS_WEBHOOK_URL first (and never commit it: the URL is a "
                    + "write credential for the channel)");
            System.exit(2);
        }

        // Share one client. The rate limiter lives on the instance, so a client per call paces
        // nothing.
        WorkflowsWebhookClient teams = WorkflowsWebhookClient.create(url);

        try {
            WebhookResponse response = teams.send(Cards.webhookCard()
                    .text("Deploy failed: api", t -> t.weight(FontWeight.BOLDER)
                            .size(FontSize.LARGE)
                            .color(Colors.ATTENTION))
                    .facts(f -> f.add("Commit", "9f2c1ab")
                            .add("Reason", "health check timed out")
                            .add("At", Instant.now().toString()))
                    .openUrl("View logs", "https://ci.example.com/builds/4711"));

            System.out.println("HTTP " + response.statusCode() + " in " + response.attempts() + " attempt(s)");
        } catch (WebhookException e) {
            // A notification is a side errand. Letting this escape would replace the failure you
            // were reporting with a failure to report it.
            System.err.println("could not notify Teams: " + e.getMessage());
        }
    }

    private DeployNotification() {}
}

File: examples/kotlin-coroutines/src/main/kotlin/example/DeployNotification.kt
package example

import io.github.teams4j.cards.Colors
import io.github.teams4j.cards.FontSize
import io.github.teams4j.cards.FontWeight
import io.github.teams4j.cards.kotlin.adaptiveCard
import io.github.teams4j.webhook.WebhookException
import io.github.teams4j.webhook.WorkflowsWebhookClient
import io.github.teams4j.webhook.kotlin.sendAwait
import java.time.Instant
import kotlinx.coroutines.runBlocking

/**
 * The same notification from a coroutine, with kotlinx.serialization and no Jackson.
 *
 * ```
 * export TEAMS_WEBHOOK_URL='https://...'
 * ./gradlew :kotlin-coroutines:run
 * ```
 *
 * Two things are worth noticing.
 *
 * `sendAwait`, not `send`. `send` is a member function, and a member always beats an extension in
 * resolution — a `suspend fun send` would be shadowed silently and the coroutine would block its
 * thread with nothing to say so. The different name is how you get the thing you meant.
 *
 * No Jackson. `teams4j-webhook` names no JSON binding, so declaring `teams4j-cards-kotlinx` is
 * enough to keep this graph Jackson-free — `./gradlew :kotlin-coroutines:check` asserts exactly
 * that.
 */
fun main(): Unit = runBlocking {
    val webhookUrl = System.getenv("TEAMS_WEBHOOK_URL")
        ?: error("set TEAMS_WEBHOOK_URL first (and never commit it: the URL is a write credential)")

    val teams = WorkflowsWebhookClient.create(webhookUrl)

    val card = adaptiveCard {
        body {
            textBlock("Deploy failed: api") {
                weight = FontWeight.BOLDER
                size = FontSize.LARGE
                color = Colors.ATTENTION
            }
            factSet {
                fact("Commit", "9f2c1ab")
                fact("Reason", "health check timed out")
                fact("At", Instant.now().toString())
            }
        }
        // `webhookActions`, not `actions`: this scope has no actionSubmit, because a webhook has no
        // bot behind it to receive one.
        webhookActions {
            actionOpenUrl("View logs", "https://ci.example.com/builds/4711")
        }
    }

    try {
        val response = teams.sendAwait(card)
        println("HTTP ${response.statusCode()} in ${response.attempts()} attempt(s)")
    } catch (e: WebhookException) {
        // Same reason as the Java example: the notification must not become the failure.
        System.err.println("could not notify Teams: ${e.message}")
    }
}
